docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+
+## Custom Param Types
+You can define your own param types using the `createParam` utility. This serves three purposes:
+1. Define the type of the param.
+2. Define how the param should be serialized and deserialized.
+3. Determine if a value is valid.
+
+Using custom param types you can define more complex validation rules. For example, this month param must be a valid month name. If it is not, the route will not match.
+
+```ts {3-11,17}
+import { createParam, createRoute, path } from '@kitbag/router'
+
+const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'] as const
+
+const monthParam = createParam((value, { invalid }) => {
+  if (months.includes(value as any)) {
+    return value
+  }
+
+  invalid(`Invalid month: ${value}`)
+})
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: monthParam,
+  }),
+})
+```
+
+By default custom params will use `toString` when serializing. But you can use a Get/Set param to define how the param should be serialized and deserialized. For example we can make the month param capitalized when parsing and lowercase when serializing.
+
+```ts
+import { createParam } from '@kitbag/router'
+
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'] as const
+
+const monthParam = createParam({
+  get: (value, { invalid }) => {
+    const month = value.charAt(0).toUpperCase() + value.slice(1)
+
+    if (months.includes(month as any)) {
+      return month
+    }
+
+    invalid(`Invalid month: ${month}`)
+  },
+  set: (value) => {
+    return value.toLowerCase()
+  },
+})
+```
+
+## Default Values
+You can define a default value for a param type when creating it. This value will be used when the param is optional and not provided in the url. An optional param of type `String` will have the type `string | undefined` when being accessed. But when a default value is provided it will have the type `string`, even if the value is missing from the url.
+
+```ts {8}
+const monthParam = createParam({
+  get: (value, { invalid }) => {
+    ...
+  },
+  set: (value) => {
+    ...
+  },
+  defaultValue: 'january',
+})
+```
+
+You can also define a default value for a individual param using the `withDefault` utility.
+
+```ts {7}
+import { createRoute, path, withDefault } from '@kitbag/router'
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: withDefault(monthParam, 'january'),
+  }),
+})
+```
+
+Or create a new param type with a default value from an existing param type.
+
+```ts {7}
+import { createParam, withDefault } from '@kitbag/router'
+
+const monthParam = createParam((value, { invalid }) => {
+  ...
+})
+
+const monthParamWithDefault = withDefault(monthParam, 'january')
+```
```

Comment:
I'd probably have skipped this too. It's neat and hopefully will bring delight to someone who tries it but probably doesn't need to be explicitly mentioned as a core-concept

Reply:
I think it helps explain how the `withDefault` utility works by creating a new param. I'd like to keep it unless you feel strongly.